lib/backend/k8s/resources/customnoderesource.go

```diff
@@ -0,0 +1,491 @@
+	node, err = c.clientSet.Nodes().Update(node)
+	if err != nil {
+		logContext.WithError(err).Info("Error updating Node when deleting resource")
+		return K8sErrorToCalico(err, kvp.Key)
+	}
+
+	return nil
+}
+
+func (c *customK8sNodeResourceClient) Get(key model.Key) (*model.KVPair, error) {
+	logContext := log.WithFields(log.Fields{
+		"Key":      key,
+		"Resource": c.resource,
+	})
+	logContext.Debug("Get per-Node resource")
+
+	// Get the names and the latest Node settings associated with the Key.
+	_, resName, node, err := c.getNamesAndNodeFromKey(key)
+	if err != nil {
+		logContext.WithError(err).Info("Error getting current settings when getting resource.")
+		return nil, err
+	}
+
+	// Extract the resource from the annotations.  It should exist.
+	kvps, err := c.extractFromAnnotations(node, resName)
+	if err != nil {
+		logContext.WithError(err).Error("Error extracting annotation when getting resource")
+		return nil, err
+	}
+	if len(kvps) != 1 {
+		logContext.Info("Resource doesn't exist")
+		return nil, errors.ErrorResourceDoesNotExist{Identifier: key}
+	}
+
+	return kvps[0], nil
+}
+
+func (c *customK8sNodeResourceClient) List(list model.ListInterface) ([]*model.KVPair, string, error) {
+	logContext := log.WithFields(log.Fields{
+		"ListInterface": list,
+		"Resource":      c.resource,
+	})
+	logContext.Debug("List per-Node Resources")
+	kvps := []*model.KVPair{}
+
+	// Extract the Node and Name from the ListInterface.
+	nodeName, resName, err := c.converter.ListInterfaceToNodeAndName(list)
+	if err != nil {
+		logContext.WithError(err).Info("Unable to determine node name")
+		return nil, "", err
+	}
+
+	// Get a list of the required nodes - which will either be all of them
+	// or a specific node.
+	var nodes []apiv1.Node
+	var rev string
+	if nodeName != "" {
+		newLogContext := logContext.WithField("NodeName", nodeName)
+		node, err := c.clientSet.Nodes().Get(nodeName, metav1.GetOptions{})
+		if err != nil {
+			newLogContext.WithError(err).Info("Error getting node")
+			err = K8sErrorToCalico(err, nodeName)
+			if _, ok := err.(errors.ErrorResourceDoesNotExist); !ok {
+				newLogContext.WithError(err).Info("Error getting node")
+				return nil, "", err
+			}
+			return kvps, "", nil
+		}
+		nodes = append(nodes, *node)
+		rev = node.GetResourceVersion()
+	} else {
+		nodeList, err := c.clientSet.Nodes().List(metav1.ListOptions{})
+		if err != nil {
+			logContext.WithError(err).Info("Error listing nodes")
+			err = K8sErrorToCalico(err, nodeName)
+			return nil, "", err
+		}
+		nodes = nodeList.Items
+		rev = nodeList.GetResourceVersion()
+	}
+
+	// Loop through each of the nodes and extract the required data.
+	for _, node := range nodes {
+		nodeKVPs, err := c.extractFromAnnotations(&node, resName)
+		if err != nil {
+			logContext.WithField("NodeName", node.GetName()).WithError(err).Warning("Error listing resources for Node")
```

Comment:
These are all *unexpected* error cases - but I think it's nice to be able to handle as gracefully as we can these cases.